schedule.py

```diff
@@ -258,89 +306,146 @@ def print_results(solver, classes, professors, sections, semesters):
+    for c in profs_classes:
         model.Add(sum(time_assign[(c, t)] for t in times) == 1)
 
     # time slots for a professor don't conflict
-    for p in sem_prof_class:
-        model.Add(
-            sum(
-                time_assign[((p, c1), t1)] * time_assign[((p, c2), t2)] * t1.conflicts[t2]
-                for c1 in sem_prof_class[p] 
-                for c2 in sem_prof_class[p] 
-                for t1 in times 
-                for t2 in times
-            ) == 1
-        )
-
-    # Creates the solver and solve.
-    solver = cp_model.CpSolver()
-    solver.Solve(model)
-    assert solver.StatusName() != 'INFEASIBLE', 'PROBLEM IS INFEASIBLE'
-    # print results
-    for c in sem_class:
+    for prof_name, sec_name in prof_teach.items():
+        for s1 in sec_name:
+            for s2 in sec_name:
+                if s1 != s2:
+                    for t1 in times:
+                        for t2 in times:
+                            if t1.conflicts[t2]:
+                                content = [time_assign[((prof_name, s1), t1)], time_assign[((prof_name, s2), t2)]]
+                                model.AddMultiplicationEquality(0, content)
+
+    # Minimize the overall time conflicts
+    conflicts = {}
+    conflict_name_template = 'course {} (timeslot {}) and course {} (timeslot {}) conflict'
+    for time1 in times:
+        for time2 in times:
+            if not time1.conflict(time2):
+                continue
+            for course1 in profs_classes:
+                for course2 in profs_classes:
+                    if course1 == course2:
+                        continue
+                    # (course1, time1) && (course2, time2) -> (course1, time1, course2, time2)
+                    key = (course1, time1, course2, time2)
+                    # create the conflict variable
+                    conflicts[key] = model.NewBoolVar(conflict_name_template.format(*key))
+                    model.Add(conflicts[key] == 1).OnlyEnforceIf([
+                        time_assign[(course1, time1)],
+                        time_assign[(course2, time2)],
+                    ])
+
+    # Maximize the number of time slots that profs prefer
+    prefer_time = {}
+    for prof_name, sec_name in prof_teach.items():
+        for s in sec_name:
+            for t in times:
+                key = (prof_name, s, t)
+                prefer_time[key] = model.NewBoolVar('{} teaches course {} on timeslots {}'.format(prof_name, s, t))
+                # model.AddHint(prefer_time[key], 0) # the value for prefer_time when c-t not scheduled?
+                model.Add(prefer_time[key] ==
+                          (professors[prof_name].prefer_time(t))).OnlyEnforceIf(
+                    [time_assign[((prof_name, s), t)]]
+                )
+
+    # equalize their importance, then multiple their weights
+    model.Minimize(3 * len(prefer_time) * sum(conflicts.values())
+                   - 2 * len(conflicts) * sum(prefer_time.values()))
```

Comment:
I don't think multiplying by the lengths of the variables is necessary - if we just have 3 and -2, it would mean that 2 time conflicts is worth 3 time preferences.

Reply:
I'm not sure if it is minimizing the variables as two sums or as individual variables. There are more variables for conflicts, which make their sum larger and potentially more important?

Reply:
I see. Even then, shouldn't it be `3 / len(...)` instead of `3 * len(...)`? If you want the total weights of conflicts and preferences to be 3:2?

Reply:
The solver only takes integers, and divisions cause errors.
I'm imagining each len()*sum() as 1 unit. Then multiply them by their weights, and make the total be 5 units. So conflicts weigh 3/5 and preferences weight 2/5.